tests/unit/test_exceptions.py

```diff
@@ -0,0 +1,395 @@
+            note="You did something wrong, which is what caused this error.",
+            hint=None,
+        )
+
+        assert err.present_ascii() == textwrap.dedent(
+            """\
+            Error: test-diagnostic
+
+            Oh no!
+            It broke. :(
+            Relevant context: Something went wrong
+                              very wrong.
+
+            Note: You did something wrong, which is what caused this error.
+            """
+        )
+
+    def test_no_context_no_hint(self) -> None:
+        err = DiagnosticPipError(
+            reference="test-diagnostic",
+            message="Oh no!\nIt broke. :(",
+            context=None,
+            note="You did something wrong, which is what caused this error.",
+            hint=None,
+        )
+
+        assert err.present_ascii() == textwrap.dedent(
+            """\
+            Error: test-diagnostic
+
+            Oh no!
+            It broke. :(
+
+            Note: You did something wrong, which is what caused this error.
+            """
+        )
+
+    def test_no_context_no_note(self) -> None:
+        err = DiagnosticPipError(
+            reference="test-diagnostic",
+            message="Oh no!\nIt broke. :(",
+            context=None,
+            note=None,
+            hint="Do it better next time, by trying harder.",
+        )
+
+        assert err.present_ascii() == textwrap.dedent(
+            """\
+            Error: test-diagnostic
+
+            Oh no!
+            It broke. :(
+
+            Hint: Do it better next time, by trying harder.
+            """
+        )
+
+    def test_no_hint_no_note(self) -> None:
+        err = DiagnosticPipError(
+            reference="test-diagnostic",
+            message="Oh no!\nIt broke. :(",
+            context="Something went wrong\nvery wrong.",
+            note=None,
+            hint=None,
+        )
+
+        assert err.present_ascii() == textwrap.dedent(
+            """\
+            Error: test-diagnostic
+
+            Oh no!
+            It broke. :(
+            Relevant context: Something went wrong
+                              very wrong.
+            """
+        )
+
+    def test_no_hint_no_note_no_context(self) -> None:
+        err = DiagnosticPipError(
+            reference="test-diagnostic",
+            message="Oh no!\nIt broke. :(",
+            context=None,
+            hint=None,
+            note=None,
+        )
+
+        assert err.present_ascii() == textwrap.dedent(
+            """\
+            Error: test-diagnostic
+
+            Oh no!
+            It broke. :(
+            """
+        )
+
+
+class TestDiagnosticPipErrorPresentation_Fancy:
+    def test_complete(self) -> None:
+        err = DiagnosticPipError(
+            reference="test-diagnostic",
+            message="Oh no!\nIt broke. :(",
+            context="Something went wrong\nvery wrong.",
+            note="You did something wrong, which is what caused this error.",
```

Comment:
The distinction between note and context is not obvious. I suspect I will have a hard time remembering what to write in which.

Reply:
Context is with the error message, and note is at the end. 🤷🏽 

Would something like `note` -> `attention` make this better?

Reply:
Done -- enforced for note and hint; and renamed them to have a `_stmt` to make it clearer that they're full sentences. The message and context should also be full sentences, but I imagine we might want them to end with paths or something like that at some point, so it's probably a reasonable idea to not check for that at runtime.

Reply:
> Would something like note -> attention make this better?

Perhaps, yes. Since it is rendered with a ⚠️, attention seems more intuitive. `note_stmt` would rather be expected to render with a :information_source: